File: src/main/java/com/morak/performancetracker/context/EndpointContextManager.java
package com.morak.performancetracker.context;

import com.morak.performancetracker.description.Descriptor;
import java.util.DoubleSummaryStatistics;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.springframework.stereotype.Component;

@Component
public class EndpointContextManager implements ContextManager {

    private final Descriptor descriptor;

    public EndpointContextManager(Descriptor descriptor) {
        this.descriptor = descriptor;
    }

    @Override
    public void afterAll(Accumulator accumulator) {
        List<Result> scopes = accumulator.getResults().entrySet()
                .stream()
                .map(it -> summarizePerScope(it.getKey(), it.getValue()))
                .collect(Collectors.toList());
        ResultComposite resultComposite = new ResultComposite(TestMetadata.ROOT, scopes);
        descriptor.describe(resultComposite);
    }

    private Result summarizePerScope(String scopeName, List<MonitorResult> results) {
        Map<String, DoubleSummaryStatistics> summary = summarize(results);
        List<Result> summaries = toResult(summary);
        TestMetadata testMetadata = new TestMetadata(scopeName, "");
        return new ResultComposite(testMetadata, summaries);
    }

    private Map<String, DoubleSummaryStatistics> summarize(List<MonitorResult> results) {
        return results.stream()
                .collect(Collectors.groupingBy(
                        MonitorResult::getName,
                        Collectors.summarizingDouble(MonitorResult::getElapsed)
                ));
    }

    private List<Result> toResult(Map<String, DoubleSummaryStatistics> summary) {
        return summary.entrySet().stream()
                .map(it -> new ResultLeaf(it.getKey(), it.getValue().getAverage()))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/morak/performancetracker/context/EndpointContextManager.kt
package com.morak.performancetracker.context

import com.morak.performancetracker.ContextType
import com.morak.performancetracker.description.Descriptor
import com.morak.performancetracker.utils.ConditionalOnPropertyContains
import com.morak.performancetracker.utils.average
import org.springframework.stereotype.Component


@Component
@ConditionalOnPropertyContains(
    value = "com.morak.performance-tracker.context.type",
    containsValue = "endpoint",
    matchIfEmpty = true
)
class EndpointContextManager(
    private val descriptor: Descriptor,
    results: List<Result>
) : ContextManager {
    private val _results: MutableList<Result> = results.toMutableList()
    val results: List<Result>
        get() = _results

    override fun afterEach(accumulator: Accumulator, testMetadata: TestMetadata) {
        val scopes = accumulator.results.entries
            .map { summarizePerScope(it.key, it.value) }
            .toList()
        _results.addAll(scopes)
    }

    override fun afterAll(accumulator: Accumulator) {
        val resultComposite = ResultComposite(TestMetadata.ROOT, results)
        descriptor.describe(resultComposite, ContextType.ENDPOINT)
    }

    private fun summarizePerScope(scopeName: String, results: List<MonitorResult>): Result {
        val summaries = summarize(results)
        val testMetadata = TestMetadata(scopeName)
        return ResultComposite(testMetadata, summaries);
    }

    private fun summarize(results: List<MonitorResult>): List<Result> {
        return results.groupBy { it.name }
            .map { (key, value) -> ResultLeaf(key, value.average { it.elapsed }) }
            .toList()
    }
}
